zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(0).unwrap_or(&None) {
+                Some(lhs_index) => {
+                    self.mutate_node(*lhs_index, |lhs_node| {
+                        let child_lazy_fees = lhs_node
+                            .total_stake()?
+                            .bmul_bdiv(remaining_lazy_fees, descendant_stake)?;
+                        lhs_node.lazy_fees =
+                            lhs_node.lazy_fees.checked_add_res(&child_lazy_fees)?;
+                        remaining_lazy_fees =
+                            remaining_lazy_fees.checked_sub_res(&child_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(1).unwrap_or(&None) {
+                Some(rhs_index) => {
+                    self.mutate_node(*rhs_index, |rhs_node| {
+                        rhs_node.lazy_fees =
+                            rhs_node.lazy_fees.checked_add_res(&remaining_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
+        }
+        self.mutate_node(index, |node| {
+            node.lazy_fees = Zero::zero();
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    fn children(&self, index: u32) -> Result<[Option<u32>; 2], DispatchError> {
+        let calculate_child =
+            |child_index: u32| Some(child_index).filter(|&i| i < self.node_count());
+        let left_child_index = index.checked_mul_res(&2)?.checked_add_res(&1)?;
+        let left_child = calculate_child(left_child_index);
+        let right_child_index = left_child_index.checked_add_res(&1)?;
+        let right_child = calculate_child(right_child_index);
+        Ok([left_child, right_child])
+    }
+
+    fn parent_index(&self, index: u32) -> Option<u32> {
+        if index == 0 {
+            None
+        } else {
+            // Won't ever fail, always returns `Some(...)`.
+            index.checked_sub(1)?.checked_div(2)
+        }
+    }
+
+    fn path_to_node(&self, mut index: u32) -> Result<Vec<u32>, DispatchError> {
+        let mut path = Vec::new();
+        let mut iterations = 0;
+        let max_iterations = Self::max_depth().checked_add_res(&1)?;
+        while let Some(parent_index) = self.parent_index(index) {
+            if iterations == max_iterations {
+                return Err(LiquidityTreeError::MaxIterationsReached.into_dispatch::<T>());
+            }
+            path.push(index);
+            index = parent_index;
+            iterations = iterations.checked_add_res(&1)?;
+        }
+        path.push(0u32); // The tree's root is not considered in the loop above.
+        path.reverse(); // The path should be from root to the node
+        Ok(path)
+    }
```

Comment:
> Assume you put in the index, which is one before the root node.

So `index = 1` or `index = 2`.

> [...] (because self.parent_index(index) will get None)

No, it will return `0`.

Reply:
nevermind, I don't know what type of a node was in my head